### PR TITLE
Support rewriting push rules when our internal defaults change

### PR DESCRIPTION
**Pull request by turt2live**
and change our internal default for tombstones

### CODE REVIEW

File: src/sync.js [hunk 1]
```diff
@@ -1030,8 +1031,9 @@ SyncApi.prototype._processSyncResponse = async function(
                 // honour push rules that were previously cached. Base rules
                 // will be updated when we recieve push rules via getPushRules
                 // (see SyncApi.prototype.sync) before syncing over the network.
-                if (accountDataEvent.getType() == 'm.push_rules') {
-                    client.pushRules = accountDataEvent.getContent();
+                if (accountDataEvent.getType() === 'm.push_rules') {
+                    const rules = accountDataEvent.getContent();
+                    client.pushRules = PushProcessor.rewriteDefaultRules(rules);
```
Comment (bwindels) [inline src/sync.js]: do we still need to do `applyRuleDefaults(client.pushRules)` in `PushProcessor::actionsForEvent` with this change? If not, we could get rid of `applyRuleDefaults`, which is very similar to the `rewriteDefaultRules` method introduced.

File: src/pushprocessor.js [hunk 2]
```diff
@@ -43,6 +43,11 @@ const DEFAULT_OVERRIDE_RULES = [
                 key: "type",
                 pattern: "m.room.tombstone",
             },
+            {
+                kind: "event_match",
```
Comment (bwindels) [inline src/pushprocessor.js]: this will match any non-state event? What is this used for?

File: src/sync.js (hunk 1 above)
Comment (turt2live) [reply src/sync.js]: They do slightly different things for different points in time. The `rewriteDefaultRules` replaces rules the server already sent to the client whereas `applyRuleDefaults` adds rules that are missing from the server. The intention is that we don't add default rules early for fear of confusing the app that the rules exist, as they are likely to try and modify them against the homeserver and not through our `PushProcessor` stuff. When they try and modify them with the homeserver, they are likely to receive errors or problems due to the name of the rule.

TLDR: we need both because they do subtly different things.
